lib/spack/spack/test/cmd_extensions.py

```diff
@@ -5,24 +5,50 @@
 
 import pytest
 
+import os
 import sys
 
+import spack.cmd
 import spack.config
+import spack.extensions
 import spack.main
 
 
-@pytest.fixture()
-def extension_root(tmpdir):
-    root = tmpdir.mkdir('spack-testcommand')
-    root.ensure('testcommand', 'cmd', dir=True)
-    return root
+class Extension:
+    def __init__(self, name, root):
+        self.name = name
+        self.pname = spack.cmd.python_name(name)
+        self.root = root
+        self.main = self.root.ensure(self.pname, dir=True)
+        self.cmd = self.main.ensure('cmd', dir=True)
 
+    def add_command(self, command_name, contents):
+        spack.cmd.require_cmd_name(command_name)
+        python_name = spack.cmd.python_name(command_name)
+        cmd = self.cmd.ensure(python_name + '.py')
+        cmd.write(contents)
 
-@pytest.fixture()
-def hello_world_cmd(extension_root):
-    """Simple extension command with code contained in a single file."""
-    hello = extension_root.ensure('testcommand', 'cmd', 'hello.py')
-    hello.write("""
+
+@pytest.fixture(params=['testcommand'], scope='function')
+def extension(request, tmpdir, config):
+    """Create a basic extension command directory structure"""
+    extension_name = request.param
+    root = tmpdir.mkdir('spack-' + extension_name)
+    extension = Extension(extension_name, root)
+
+    list_of_modules = list(sys.modules.keys())
+    with spack.config.override('config:extensions', [str(extension.root)]):
+        yield extension
+
+    to_be_deleted = [x for x in sys.modules if x not in list_of_modules]
+    for module_name in to_be_deleted:
+        del sys.modules[module_name]
```

Comment:
@alalazo I might need some more detail to understand how that would work with cleanup actions. I agree that the use of `indirect` adds some complexity. It is documented at http://doc.pytest.org/en/latest/example/parametrize.html#apply-indirect-on-particular-arguments and IMO that is reasonably clear, so I don't think this is a major issue. One issue I have with this approach is that it identifies the name of the fixture as a string, so any future typo could lead to strange and hard-to-debug errors.

Reply:
@scheibelp The cleanup action for modules would stay the same as here. My thought was that instead of having `request` here as a fixture, we would just create a closure taking `extension_name` as argument and yield that. At that point one can just parametrize on extension names in single tests and use this fixture to create the extension with one single call.

Reply:
I attempted this, but there are complications I haven't been able to surmount -- the indirect nature means having to pass the name through at least one other fixture to get it to `create_extension()`, and I haven't been able to find out how to deal with the `with...` clause in the generator function. If you have some pointers for me I could try again, but to be honest it seems like a lot of work just to avoid one `indirect=True` parameter.

Reply:
@alalazo @scheibelp I believe I have something that works with 7cc2c8c71. I'll leave it up to you to decide if it's an improvement.

Reply:
This change is what I had in mind for the fixture, so if it's fine with @scheibelp it's fine with me.